### PR TITLE
[12.0][IMP] account_credit_control: preventive policy.

### PR DESCRIPTION
**Pull request by ernestotejeda**
Negative numbers can now be specified in policy levels.

cc @Tecnativa TT24674

### CODE REVIEW

**Comment by rafaelbn** [comment]
@ernestotejeda I found and error. Maybe is not related to this PR but we should review it.

Go to Credit Control Run -> Compute credit control lines and you get:

Error:
Odoo Server Error

Traceback (most recent call last):
  File "/.repo_requirements/odoo/odoo/api.py", line 1049, in get
    value = self._data[key][field][record._ids[0]]
KeyError: 2

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/.repo_requirements/odoo/odoo/fields.py", line 1012, in __get__
    value = record.env.cache.get(record, self)
  File "/.repo_requirements/odoo/odoo/api.py", line 1051, in get
    raise CacheMiss(record, field)
odoo.exceptions.CacheMiss: ('credit.control.policy(2,).policy_sign', None)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/.repo_requirements/odoo/odoo/models.py", line 4746, in ensure_one
    _id, = self._ids
ValueError: too many values to unpack (expected 1)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/.repo_requirements/odoo/odoo/http.py", line 656, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/.repo_requirements/odoo/odoo/http.py", line 314, in _handle_exception
    raise pycompat.reraise(type(exception), exception, sys.exc_info()[2])
  File "/.repo_requirements/odoo/odoo/tools/pycompat.py", line 87, in reraise
    raise value
  File "/.repo_requirements/odoo/odoo/http.py", line 698, in dispatch
    result = self._call_function(**self.params)
  File "/.repo_requirements/odoo/odoo/http.py", line 346, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/.repo_requirements/odoo/odoo/service/model.py", line 98, in wrapper
    return f(dbname, *args, **kwargs)
  File "/.repo_requirements/odoo/odoo/http.py", line 339, in checked_call
    result = self.endpoint(*a, **kw)
  File "/.repo_requirements/odoo/odoo/http.py", line 941, in __call__
    return self.method(*args, **kw)
  File "/.repo_requirements/odoo/odoo/http.py", line 519, in response_wrap
    response = f(*args, **kw)
  File "/home/odoo/OCB-12.0/addons/web/controllers/main.py", line 966, in call_button
    action = self._call_kw(model, method, args, {})
  File "/home/odoo/OCB-12.0/addons/web/controllers/main.py", line 954, in _call_kw
    return call_kw(request.env[model], method, args, kwargs)
  File "/.repo_requirements/odoo/odoo/api.py", line 759, in call_kw
    return _call_kw_multi(method, model, args, kwargs)
  File "/.repo_requirements/odoo/odoo/api.py", line 746, in _call_kw_multi
    result = method(recs, *args, **kwargs)
  File "/home/odoo/build/OCA/credit-control/account_credit_control/models/credit_control_run.py", line 168, in generate_credit_lines
    self._generate_credit_lines()
  File "/home/odoo/build/OCA/credit-control/account_credit_control/models/credit_control_run.py", line 138, in _generate_credit_lines
    ) = policy._generate_credit_lines(self.date, {'run_id': self.id})
  File "/home/odoo/build/OCA/credit-control/account_credit_control/models/credit_control_policy.py", line 236, in _generate_credit_lines
    lines = self._get_move_lines_to_process(controlling_date)
  File "/home/odoo/build/OCA/credit-control/account_credit_control/models/credit_control_policy.py", line 188, in _get_move_lines_to_process
    lines = self._due_move_lines(controlling_date)
  File "/home/odoo/build/OCA/credit-control/account_credit_control/models/credit_control_policy.py", line 102, in _due_move_lines
    domain_line = self._move_lines_domain(controlling_date)
  File "/home/odoo/build/OCA/credit-control/account_credit_control/models/credit_control_policy.py", line 78, in _move_lines_domain
    if self.policy_sign == "negative":
  File "/.repo_requirements/odoo/odoo/fields.py", line 1016, in __get__
    self.determine_value(record)
  File "/.repo_requirements/odoo/odoo/fields.py", line 1129, in determine_value
    self.compute_value(recs)
  File "/.repo_requirements/odoo/odoo/fields.py", line 1083, in compute_value
    self._compute_value(records)
  File "/.repo_requirements/odoo/odoo/fields.py", line 1074, in _compute_value
    getattr(records, self.compute)()
  File "/home/odoo/build/OCA/credit-control/account_credit_control/models/credit_control_policy.py", line 61, in _compute_policy_sign
    for level in self.level_ids:
  File "/.repo_requirements/odoo/odoo/fields.py", line 1010, in __get__
    record.ensure_one()
  File "/.repo_requirements/odoo/odoo/models.py", line 4749, in ensure_one
    raise ValueError("Expected singleton: %s" % self)
ValueError: Expected singleton: credit.control.policy(2, 1, 3)

File: account_credit_control/models/credit_control_policy.py [hunk 1]
```diff
@@ -44,17 +45,41 @@ class CreditControlPolicy(models.Model):
...
+        ],
+        compute='_compute_policy_sign',
+    )
+
+    @api.depends('level_ids')
+    def _compute_policy_sign(self):
+        mapping = {-1: 'negative', 0: False, 1: 'positive'}
+        for record in self:
+            policy_sign = 0
+            for level in self.level_ids:
```
Comment (pedrobaeza) [inline account_credit_control/models/credit_control_policy.py]:
```suggestion
            for level in record.level_ids:
```

**Review by pedrobaeza — CHANGES_REQUESTED** [review]
@rafaelbn the error is here in this PR.

**Comment by ernestotejeda** [comment]
Fixed.
@rafaelbn @pedrobaeza could you review?, please.

**Review by pedrobaeza — APPROVED** [review]
Code review

**Review by marcelsavegnago — APPROVED** [review]

**Review by rafaelbn — APPROVED** [review]
:+1:

**Comment by pedrobaeza** [comment]
/ocabot merge major

@ernestotejeda please migrate it to v13 as well.

**Pull request merged**